crypto/bio/bio_meth.c

```diff
@@ -63,12 +63,45 @@ int BIO_meth_set_write(BIO_METHOD *biom,
 
 int (*BIO_meth_get_read(BIO_METHOD *biom)) (BIO *, char *, int)
 {
+    return biom->bread_old;
+}
+
+int (*BIO_meth_get_read_ex(BIO_METHOD *biom)) (BIO *, char *, size_t, size_t *)
+{
     return biom->bread;
 }
 
+/* Conversion for old style bread to new style */
+int bread_conv(BIO *bio, char *out, size_t outl, size_t *read)
+{
+    int ret;
+
+    if (outl > INT_MAX)
+        return 0;
+
+    ret = bio->method->bread_old(bio, out, (int)outl);
+
+    if (ret <= 0) {
+        *read = 0;
+        return ret;
+    }
```

Comment:
If there is data to be read prior to reaching eof then as much data as we can get will be read into the buffer, and then a success return will be given. If you attempt to read again then there is no data to be read and it returns as a failure. This is the same as fread, and the same as now for existing BIO_read().